### ISSUE TITLE
rmse in R package appears to be incorrect

### ISSUE THREAD

**Issue by zachmayer** [dmlc/mxnet#1773]
If I compare RMSE on a validation set, as reported by mxnet during training, to a by-hand calculation

```{R}
require(mlbench)
require(mxnet)
data(BostonHousing, package="mlbench")

train.ind <- seq(1, 506, 3)
train.x <- data.matrix(BostonHousing[train.ind, -14])
train.y <- BostonHousing[train.ind, 14]
test.x <- data.matrix(BostonHousing[-train.ind, -14])
test.y <- BostonHousing[-train.ind, 14]

data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)

mx.callback.trueerror <- function(period=1) {
  function(iteration, nbatch, env, verbose) {
    if (iteration %% period == 0) {
      p1 <- predict(env$model, test.x, array.layout='rowmajor')[1,]
      if(verbose) cat(paste("True Validation-rmse:", sqrt(mean((p1 - test.y)^2))))
    }
    return(TRUE)
  }
}

mx.set.seed(0)
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  ctx=mx.cpu(), num.round=50, array.batch.size=20,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse,
  eval.data = list(data=test.x, label=test.y),
  epoch.end.callback=mx.callback.trueerror(1)
  )
```

Yields the following output:
```{R}
[50] Train-rmse=8.24580523984314
[50] Validation-rmse=7.632446135899
True Validation-rmse: 7.80050228847714
````

The reported RMSE from mxnet is different than what I get when I calculate by hand.  What am I doing wrong here?

Thanks,

Zach

**Comment by gpierard** [comment]
Hey, having the same problem here, see my post [https://github.com/dmlc/mxnet/issues/1812](url)

**Comment by thirdwing** [comment]
Hi, @zachmayer really sorry for this late reply.

I think you are correct and mxnet is also correct.

During the training, the rmse is calculated in a batch and take average. In your case, the size of training and validation set is not divisible by the batch size, so the result might be slightly different.

Can you try the code below? I just change the batch size and now you should get consistent rsme.

```r
require(mlbench)
require(mxnet)
data(BostonHousing, package="mlbench")

train.ind <- sample(1:506, 253)
train.x <- data.matrix(BostonHousing[train.ind, -14])
train.y <- BostonHousing[train.ind, 14]
test.x <- data.matrix(BostonHousing[-train.ind, -14])
test.y <- BostonHousing[-train.ind, 14]

data <- mx.symbol.Variable("data")
fc1 <- mx.symbol.FullyConnected(data, num_hidden=1)
lro <- mx.symbol.LinearRegressionOutput(fc1)

mx.callback.trueerror <- function(period=1) {
  function(iteration, nbatch, env, verbose) {
    if (iteration %% period == 0) {
      p1 <- predict(env$model, test.x, array.layout='rowmajor')[1,]
      if(verbose) cat(paste("True Validation-rmse:", sqrt(mean((p1 - test.y)^2))))
    }
    return(TRUE)
  }
}

mx.set.seed(0)
model <- mx.model.FeedForward.create(
  lro, X=train.x, y=train.y,
  ctx=mx.cpu(), num.round=50, array.batch.size=253,
  learning.rate=2e-6, momentum=0.9, eval.metric=mx.metric.rmse,
  eval.data = list(data=test.x, label=test.y),
  epoch.end.callback=mx.callback.trueerror(1)
)
```

**Comment by zachmayer** [comment]
Thanks, I'll take a look!

**Comment by thirdwing** [comment]
Close this for now. Feel free to reopen.
